### PR TITLE
Support RSA and x509 formats

### PR DESCRIPTION
**Pull request by EndrII**
# Work plan

## Implementations:

- [x] ESDSA SSL 3.0
- [x] RSA SSL 3.0
- [x] X509
- [x] Documentation

## Tests 

- [x] ESDSA SSL 3.0
- [x] RSA SSL 3.0
- [x] X509

### CODE REVIEW

File: tests/tstMain.cpp [hunk 1]
```diff
@@ -32,12 +38,14 @@ class tstMain : public QObject
...
     // BEGIN TESTS CASES
     TestCase(authTest, AuthTest)
-    TestCase(cryptoTest, CryptoTest<EasySSL::ECDSASSL11>)
+    TestCase(cryptoTestESDSA, CryptoTest<EasySSL::ECDSASSL>)
+    TestCase(cryptoTestRSA, CryptoTest<EasySSL::RSASSL>)
+    TestCase(crtTestX509RSA, CrtTestX509RSA)
+    TestCase(crtTestX509ECDSA, CrtTestX509ECDSA)
 
-    // END TEST CASES
+    // END TEST CASESa
```
Comment (EndrII) [inline tests/tstMain.cpp]:
```suggestion
    // END TEST CASES
```

File: src/lib/src/private/easysslutils.h [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+//# of this license document, but changing it is not allowed.
+//#
+
+#include <openssl/types.h>
+
+#include <QByteArray>
+namespace EasySSL {
+
+/**
+ * @brief The EasySSLUtils class This is base utils for work with opwnssl library.
```
Comment (usermeme) [inline src/lib/src/private/easysslutils.h]:
```suggestion
 * @brief The EasySSLUtils class These are basic utils for work with the opwnssl library.
```

File: src/lib/src/private/easysslutils.h [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+ */
+class EasySSLUtils {
+
+public:
+
+    /**
+     * @brief printlastOpenSSlError This method print last ssl error message.
```
Comment (usermeme) [inline src/lib/src/private/easysslutils.h]:
```suggestion
     * @brief printlastOpenSSlError This method prints the latest ssl error message.
```

File: src/lib/src/private/easysslutils.h [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+     */
+    static void printlastOpenSSlError();
+
+    /**
+     * @brief bignumToArray This method convert openssl BIGNUM into byteArray
```
Comment (usermeme) [inline src/lib/src/private/easysslutils.h]:
```suggestion
     * @brief bignumToArray This method converts openssl BIGNUM into byteArray
```

File: src/lib/src/private/easysslutils.h [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+     * @param num This is big num of the openssl  library
```
Comment (usermeme) [inline src/lib/src/private/easysslutils.h]:
```suggestion
     * @param num This is a big num of the openssl library
```

File: src/lib/src/private/easysslutils.h [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+     * @return bytes array.
+     */
+    static QByteArray bignumToArray(const BIGNUM* num);
+
+    /**
+     * @brief bignumFromArray This method convert Qt bytes array into opensll big num.
```
Comment (usermeme) [inline src/lib/src/private/easysslutils.h]:
```suggestion
     * @brief bignumFromArray This method converts the Qt bytes array into the opensll big num.
```

File: src/lib/src/private/easysslutils.h [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+     * @param array This is input array.
```
Comment (usermeme) [inline src/lib/src/private/easysslutils.h]:
```suggestion
     * @param array This is an input array.
```

File: src/lib/src/private/easysslutils.h [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+     * @return big num pointer.
+     * @note This result pointer will not free automatically. Please free returned pointer after using.
```
Comment (usermeme) [inline src/lib/src/private/easysslutils.h]:
```suggestion
     * @note This result pointer will not automatically be free. Please free the returned pointer after use.
```

File: src/lib/src/private/easysslutils.h [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+     */
+    [[nodiscard("This pointer will not free automatically. Please free returned pointer after using.")]]
```
Comment (usermeme) [inline src/lib/src/private/easysslutils.h]:
```suggestion
    [[nodiscard("This result pointer will not automatically be free. Please free the returned pointer after use.")]]
```

File: src/lib/src/private/easysslutils.h [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+    static BIGNUM* bignumFromArray(const QByteArray& array);
+
+    /**
+     * @brief bioToByteArray This method conver openssl BIO to QByteArry
```
Comment (usermeme) [inline src/lib/src/private/easysslutils.h]:
```suggestion
     * @brief bioToByteArray This method converts the openssl BIO to the QByteArry
```

File: src/lib/src/private/easysslutils.h [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+     * @param bio input arrary.
+     * @return Qt Array
+     */
+    static QByteArray bioToByteArray(BIO *bio);
+
+    /**
+     * @brief byteArrayToBio This method create BIO struct from the Qt QByteArray object.
+     * @param byteArray This is input Qt byte array.
```
Comment (usermeme) [inline src/lib/src/private/easysslutils.h]:
```suggestion
     * @brief byteArrayToBio This method creates the BIO struct from the Qt QByteArray object.
     * @param byteArray This is an input Qt byte array.
```

File: src/lib/src/private/easysslutils.h [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+     * @return pointer tot the BIO.
```
Comment (usermeme) [inline src/lib/src/private/easysslutils.h]: what is the "tot"

File: src/lib/src/private/easysslutils.h [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+     * @note Do not forget free result pointer.
```
Comment (usermeme) [inline src/lib/src/private/easysslutils.h]:
```suggestion
     * @note Don't forget to clear the result pointer.
```

File: src/lib/src/private/easysslutils.h [hunk 14]
```diff
@@ -0,0 +1,75 @@
...
+     */
+    [[nodiscard("This pointer will not free automatically. Please free returned pointer after using.")]]
+    static BIO *byteArrayToBio(const QByteArray &byteArray);
+
+    /**
+     * @brief extractPublcKey This method extracts publick key from the ssl (pem) structure.
```
Comment (usermeme) [inline src/lib/src/private/easysslutils.h]:
```suggestion
     * @brief extractPublcKey This method extracts the public key from the ssl (pem) structure.
```

File: src/lib/src/private/easysslutils.h [hunk 15]
```diff
@@ -0,0 +1,75 @@
...
+     * @param ssl_keys This is ssl keys objects.
```
Comment (usermeme) [inline src/lib/src/private/easysslutils.h]:
```suggestion
     * @param ssl_keys These are objects of the ssl keys.
```

File: src/lib/src/private/easysslutils.h [hunk 16]
```diff
@@ -0,0 +1,75 @@
...
+     * @return bytes array of the extracted key.
+     */
+    static QByteArray extractPublcKey(EVP_PKEY* ssl_keys);
+
+    /**
+     * @brief extractPrivateKey This method extracts private key from the ssl (pem) structure.
```
Comment (usermeme) [inline src/lib/src/private/easysslutils.h]:
```suggestion
     * @brief extractPrivateKey This method extracts the private key from the ssl (pem) structure.
```

File: src/lib/src/private/easysslutils.h [hunk 17]
```diff
@@ -0,0 +1,75 @@
...
+     * @param ssl_keys This is ssl keys objects.
```
Comment (usermeme) [inline src/lib/src/private/easysslutils.h]:
```suggestion
     * @param ssl_keys These are objects of the ssl keys.
```

File: src/lib/src/public/easyssl/ecdsassl.h [hunk 18]
```diff
@@ -0,0 +1,80 @@
...
+#ifndef QH_ECDSA_SSL_1_1_H
+#define QH_ECDSA_SSL_1_1_H
+
+#include "global.h"
+#include "icrypto.h"
+
+namespace EasySSL {
+
+/**
+ * @brief The ECDSASSL class is ecdsa implementation of the Async authentication. This implementation based on Openssl library.
```
Comment (usermeme) [inline src/lib/src/public/easyssl/ecdsassl.h:18]:
```suggestion
 * @brief The ECDSASSL class is ecdsa implementation of the Async authentication. This implementation based on an Openssl library.
```

File: src/lib/src/public/easyssl/ecdsassl.h [hunk 19]
```diff
@@ -0,0 +1,80 @@
...
+    enum EllipticCurveStandart {
+        /// Private key (point on Elliptic Curve ) based on 256 bit prime number
+        P_256,
+        /// Private key (point on Elliptic Curve ) based on 384 bit prime number
+        P_384,
+        /// Private key (point on Elliptic Curve ) based on 521 bit prime number
+        P_521,
+        /// based on elliptic curve potentially offering 224 bits of security and designed for use with the elliptic-curve Diffie–Hellman (ECDH) key agreement scheme
+        X448,
+        ///  base on an elliptic curve used in elliptic-curve cryptography (ECC) offering 128 bits of security (256-bit key size) and designed for use with the elliptic curve Diffie–Hellman (ECDH) key agreement scheme. It is one of the fastest curves in ECC, and is not covered by any known patents.
```
Comment (usermeme) [inline src/lib/src/public/easyssl/ecdsassl.h:34]:
```suggestion
        ///  It based on an elliptic curve. It has used in the elliptic-curve cryptography (ECC). It is offering 128 bits of security (256-bit key size) and designed for using with the elliptic curve Diffie–Hellman (ECDH) key agreement scheme. It is one of the fastest curves in ECC and is not covered by any known patents.
```

File: src/lib/src/public/easyssl/ecdsassl.h [hunk 20]
```diff
@@ -0,0 +1,80 @@
...
+     * @brief The EllipticCurveStandart enum List of supported Elliptic Curve Standarts
+     */
+    enum EllipticCurveStandart {
+        /// Private key (point on Elliptic Curve ) based on 256 bit prime number
+        P_256,
+        /// Private key (point on Elliptic Curve ) based on 384 bit prime number
+        P_384,
+        /// Private key (point on Elliptic Curve ) based on 521 bit prime number
+        P_521,
+        /// based on elliptic curve potentially offering 224 bits of security and designed for use with the elliptic-curve Diffie–Hellman (ECDH) key agreement scheme
```
Comment (usermeme) [inline src/lib/src/public/easyssl/ecdsassl.h:32]:
```suggestion
        /// It based on the elliptic curve. Is is potentially offering 224 bits of security. It designed for using with the elliptic-curve Diffie–Hellman (ECDH) key agreement scheme
```

File: src/lib/src/public/easyssl/ecdsassl.h [hunk 21]
```diff
@@ -0,0 +1,80 @@
...
+    QByteArray decrypt(const QByteArray &message, const QByteArray &key) override;
+
+    /**
+     * @brief encrypt This method has empty implementation.
+     * @return empty array.
+     */
+    QByteArray encrypt(const QByteArray &message, const QByteArray &key) override;
+
+    /**
+     * @brief curve This method return current curve method. using only for generate new pair keys.
```
Comment (usermeme) [inline src/lib/src/public/easyssl/ecdsassl.h]:
```suggestion
     * @brief curve This method returns the current curve method. Using only for generating new pair of keys.
```

File: src/lib/src/private/easysslutils.h (hunk 8 above)
Comment (EndrII) [reply src/lib/src/private/easysslutils.h]:
```suggestion
     * @note This result pointer will not be free automatically. Please free the returned pointer after use.
```

File: src/lib/src/private/easysslutils.h (hunk 9 above)
Comment (EndrII) [reply src/lib/src/private/easysslutils.h]:
```suggestion
    [[nodiscard("The result pointer will not be free automatically. Please free the returned pointer after using.")]]
```

File: src/lib/src/private/easysslutils.h (hunk 12 above)
Comment (EndrII) [reply src/lib/src/private/easysslutils.h]: to
```suggestion
     * @return pointer to the BIO struct of OpenSLL library.
```

File: src/lib/src/private/easysslutils.h (hunk 13 above)
Comment (EndrII) [reply src/lib/src/private/easysslutils.h]:
```suggestion
     * @note Don't forget to free the result pointer.
```

File: src/lib/src/public/easyssl/ecdsassl.h (hunk 18 above)
Comment (EndrII) [reply src/lib/src/public/easyssl/ecdsassl.h:18]: OpenSSL, it is name of library. So 'an' is not needed on this case

**Pull request merged**